client/src/main/java/org/eclipse/hono/client/impl/HonoConnectionImpl.java

```diff
@@ -370,6 +422,9 @@ private void connect(
             final Handler<AsyncResult<HonoConnection>> connectionHandler,
             final Handler<ProtonConnection> disconnectHandler) {
 
+        // Note: in all places that the connectionHandler is completed here, the removeAndSucceedCheckConnectedPromises()
+        // or removeAndFailCheckConnectedPromises() method has to be invoked as well (completing concurrently created checkConnected() futures)!
+
```

Comment:
what about wrapping the given connectionHandler into a handler that does both. That way we can simply continue to complete one handler only ...

Reply:
Good point. I initially thought of doing that as well, but refrained from doing so because `connect()` is invoked recursively and I didn't want to decorate/wrap the handler multiple times. But yes, an explicit handler class and an `instanceof` check before wrapping the handler can help.

Reply:
> ... because connect() is invoked recursively ...

is it? I think that we invoke it repeatedly by means of scheduling a new handler on the event loop. But AFAIK we do not invoke it recursively. At least I hope that we are not doing that ...

Reply:
I meant recursively in the sense of the logical call hierarchy (ignoring Vert.x handlers in between), concentrating on the fact that the `connectionHandler` param given to the `connect()` method (and possibly being decorated/wrapped there), possibly ends up in another `connect()` invocation.